plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUClass.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiIdentifier
+import org.jetbrains.kotlin.asJava.classes.KtLightClass
+import org.jetbrains.kotlin.fir.declarations.FirClass
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getOrBuildFirOfType
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getResolveState
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.uast.*
+import org.jetbrains.uast.internal.acceptList
+import org.jetbrains.uast.visitor.UastVisitor
+
+sealed class AbstractFirKotlinUClass(
+    givenParent: UElement?
+) : FirKotlinAbstractUElement(givenParent), UClass, UAnchorOwner {
+    override val uAnnotations: List<UAnnotation>
+        get() {
+            // TODO: Not yet implemented
+            return emptyList()
+        }
+
+    override val uastDeclarations: List<UDeclaration> by lz {
+        mutableListOf<UDeclaration>().apply {
+            addAll(fields)
+            addAll(initializers)
+            addAll(methods)
+            addAll(innerClasses)
+        }
+    }
+
+    open val ktClass: KtClassOrObject?
+        get() = (psi as? KtLightClass)?.kotlinOrigin
+
+    // TODO: it's not yet clear whether this is needed
+    protected val firClass: FirClass<*>? by lz {
+        ktClass?.let {
+            it.getOrBuildFirOfType<FirClass<*>>(it.getResolveState())
+        }
+    }
+
+    override val uastSuperTypes: List<UTypeReferenceExpression>
+        get() = ktClass?.superTypeListEntries.orEmpty().mapNotNull { it.typeReference }.map {
+            FirKotlinUTypeReferenceExpression(it, this)
+        }
+
+    // TODO: delegateExpressions
+
+    override fun accept(visitor: UastVisitor) {
+        if (visitor.visitClass(this)) return
+        // TODO: delegate expressions
+        uAnnotations.acceptList(visitor)
+        uastDeclarations.acceptList(visitor)
+        visitor.afterVisitClass(this)
+    }
+
+    companion object {
+        fun create(psi: KtLightClass, givenParent: UElement?): UClass = when (psi) {
+            // TODO: PsiAnonymousClass
+            // TODO: KtLightClassForScript
+            else -> FirKotlinUClass(psi, givenParent)
+        }
+    }
+}
+
+class FirKotlinUClass(
+    psi: KtLightClass,
```

Comment:
Should we still require `KtLightClass` to create `UClass`? Shouldn't `FirClass` be there instead?

Reply:
Let's use FIR internals as less as possible, so we can use `KtClassOrObject` symbol here if we don't want to use light classes

Reply:
Yeah, I was about to answer that it's intentional _not_ to use `FirElement` at the constructor level at all.

Reply:
But probably using `KtLightClass` at the consturctor level has similar downsides as `FirElement`. So I also vote for `KtClassOrObject` then

Reply:
I played a little bit, and found it's inevitable to have a parameter with `KtLightClass` because `UClass.javaPsi` is of `PsiClass` type, I meant, non-null type. From `KtClassOrObject`, we could convert it to `KtLightClass?`, but it's not guaranteed to be non-null? I found that, before creating `KotlinUClass`, the conversion is stopped if the corresponding light-class isn't available: https://github.com/JetBrains/kotlin/blob/master/plugins/uast-kotlin/src/org/jetbrains/uast/kotlin/KotlinUastLanguagePlugin.kt#L681 and I wonder we can't do that in the constructor. I tweak/clean up creation of declarations a little bit, though.

Reply:
As comment [here](https://github.com/JetBrains/kotlin/blob/e4870740e1f1c263f432fba5c83019ca318e31d0/compiler/light-classes/src/org/jetbrains/kotlin/asJava/lightClassUtils.kt#L42) says, `null` will be returned only for non-jvm class or for script, which is not a case for UAST

Reply:
Actually, we have a three options here:
- store psi in UClass: `KtClassOrObject`, then we have to get light class and `KtClassOrObjectSymbol` by it;
- store KtSymbol in UClass: `KtClassOrObjectSymbol` by which we can get a light class and `KtClassOrObject`;
- store light class which contains `KtClassOrObjectSymbol` inside and it's possible to get  `KtClassOrObject` psi by it.

I would stick with the lite-class as it already contains all info we need. And from the light class we can always get `KtClassOrObjectSymbol` as it is just stored inside. This approach may also solve the problem of getting the same symbols from different `KtAnalasysSession`. This problem is solved in light classes by sticking to the single analysis session which lived in KtSymbol

Reply:
So, it looks like we're getting back to my original approach. :) Well, not exactly same as before, but I think it's better to override `javaPsi` at the constructor, if the given parameter is used to override that anyway.

I thought `sourcePsi` is available (as non-null) while creating a light class, but perhaps not the case for enum entry? So, I took back that addition for now.